src/main/java/org/elasticsearch/common/geo/GeoPoint.java

```diff
@@ -79,38 +82,46 @@ public GeoPoint resetFromGeoHash(String hash) {
     }
 
     public final double lat() {
-        return this.lat;
+        return this.y;
     }
 
     public final double getLat() {
-        return this.lat;
+        return this.y;
     }
 
     public final double lon() {
-        return this.lon;
+        return this.x;
     }
 
     public final double getLon() {
-        return this.lon;
+        return this.x;
     }
 
     public final String geohash() {
-        return GeoHashUtils.encode(lat, lon);
+        return GeoHashUtils.encode(y, x);
     }
 
     public final String getGeohash() {
-        return GeoHashUtils.encode(lat, lon);
+        return GeoHashUtils.encode(y, x);
     }
 
     @Override
     public boolean equals(Object o) {
         if (this == o) return true;
-        if (o == null || getClass() != o.getClass()) return false;
+        if (o == null) return false;
+        if (o instanceof Coordinate) {
```

Comment:
Won't this be `true` for any `GeoPoint`, so you can never reach the block casting to `GeoPoint`?